javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -241,13 +248,16 @@ private Vector(T[] leading, byte leadingOffset, VectorTree<T> middle, T[] traili
         }
 
         final T[] leading = (T[]) copyOfRange(array, 0, Math.min(size, branchingFactor()));
-        final T[] trailing = (T[]) copyOfRange(array, Math.max(leading.length, array.length - branchingFactor()), array.length);
-        if (trailing.length < branchingFactor()) {
+        final int remaining = size - leading.length;
+        int trailingSize = lastDigit(remaining);
+        if (trailingSize == 0) { trailingSize += branchingFactor(); }
+        final T[] trailing = (T[]) copyOfRange(array, array.length - trailingSize, array.length);
+        final int middleSize = remaining - trailingSize;
```

Comment:
we can calculate the middle size to be full always, and put the rest in `trailing`